### DISCUSSION TITLE
Type annotations

### DISCUSSION THREAD

**Discussion by einarwar** [uktrade/stream-zip#120] [Ideas]
Hi there.

I came across this package when trying to zip large files in FastAPI on systems with restrictions on memory usage, and this package was just what i needed!

I noticed that there were no type-annotations, which made implementation slightly more difficult. To make use of this package easier, it could be beneficial to add type-annotations, at least for public functions. If this is something you think is good, I would be more than happy to provide a PR :)

**Comment by michalc** [comment]
Hi @einarwar,

Sounds great! If you're happy to raise a PR, that would be amazing.

In my head then in CI at https://github.com/uktrade/stream-zip/blob/main/.github/workflows/test.yml we would also run mypy on the code, and I guess the tests as well, and that should hopefully be enough to check the type annotations?

**Reply by einarwar** [reply]
Yeah that sounds like a good idea. I personally used Pyright on my other projects, but there should be no problem using mypy, as they are pretty equal.

One other thing that i noticed is that python versions from 3.6.7 and above is supported. 3.6 and 3.7 have reached EOL. I understand that this can be a breaking change for some users, but do you have any plans on requiring e.g python 3.8?

**Reply by michalc** [reply]
> Yeah that sounds like a good idea. I personally used Pyright on my other projects, but there should be no problem using mypy, as they are pretty equal.

Ah I have to admit ignorance really... not sure if I have enough knowledge to have a preference. It was more just to run _something_ to get some confidence the types make sense.

> One other thing that i noticed is that python versions from 3.6.7 and above is supported. 3.6 and 3.7 have reached EOL. I understand that this can be a breaking change for some users, but do you have any plans on requiring e.g python 3.8?

No plans on requiring Python 3.8. Even though they are EOL, I would prefer to keep support for 3.6 and 3.7 long as we can really. I'm not sure where the data comes from, but https://w3techs.com/technologies/history_details/pl-python/3 lists Python 3.6 as the most popular version of Python used in websites for example.

I wasn't aware this would be a breaking change in this respect - why would it be?

**Reply by einarwar** [reply]
I understand. I meant a breaking change in the sense that if one would require a python version of 3.8 or greater, users of this package currently running a python version less than 3.8 would not be able to get future updates to this package without migrating their projects to 3.8 as well. 

The reason i asked is that `Protocol` was added in 3.8, which i think would be a nice way of typing some of the methods used in the library. But i respect the need to support 3.6 and 3.7

**Reply by michalc** [reply]
Ah… if we wanted to use `Iterable[bytes]` would that be a Python 3.8/Protocol thing? What would be the pre-3.8 alternatives, if any? Could we even have different types for different Python versions?

**Reply by michalc** [reply]
Had some time so made a very small start on this: https://github.com/uktrade/stream-zip/pull/121

(Happy for more PRs - suspect it might take a few to say, get it to the point of strict type checking?)

**Reply by michalc** [reply]
Next step: https://github.com/uktrade/stream-zip/pull/122

**Reply by einarwar** [reply]
> Ah… if we wanted to use `Iterable[bytes]` would that be a Python 3.8/Protocol thing? What would be the pre-3.8 alternatives, if any? Could we even have different types for different Python versions?

`Iterable` and the like should be supported since python 3.6 i believe. I was think of `typing.Protocol` https://peps.python.org/pep-0544/ . Since objects/classes are passed around quite a bit in the code, `Protocols` can provide type hints to which attributes or methods the object needs to implement to be valid.
For a concrete code example: https://github.com/uktrade/stream-zip/blob/d9d79b4057cc89b13b92c728b9b9028663f4c17c/stream_zip.py#L327-L339

The `_get_compress_obj` could be typed as a `Callable` that returns a `Protocol` that implements the `compress` method. 

> (Happy for more PRs - suspect it might take a few to say, get it to the point of strict type checking?)

Having been through the process of typing an entire project that was previously untyped, my experience is that it is a long and iterative process. It has however saved me from several potential bugs, and i believe it is well worth it :)

**Reply by michalc** [reply]
> The _get_compress_obj could be typed as a Callable that returns a Protocol that implements the compress method.

Ah I see! Although am I right that it would also have to have a "flush" method?

I have now made it the "nominal" (if that's the right word) type of 'zlib._Compress'. But to check really, is there much of a use case to use a Protocol here? Or is it sort of the principal of the thing - nothing in the code at runtime really requires a `zlib._Compress` - someone should be free to use another compressor if they want, so it's too strict to enforce it at the type level?

**Reply by michalc** [reply]
In any case, have now a, hopefully, type hinted stream-zip package (as of v0.0.80 really). It should cause mypy to up with errors like the following if you pass stream_zip the wrong thing for example:

```
Argument 1 to "stream_zip" has incompatible type "int"; expected "Iterable[tuple[str, datetime, int, Method, Iterable[bytes]]]
```

**Reply by einarwar** [reply]
Wow, I am really impressed at what you did during the weekend. I was hoping to be able to contribute, but you were too fast 😅. This is really nice, thank you so much!

I used the package for something similar to the example mentioned in https://github.com/uktrade/stream-zip/discussions/31 , and with a fully typed external API (and internal for that matter) things work really smooth :)
